### PR TITLE
Adds metadata implementation

### PR DESCRIPTION
**Pull request by grnydawn**
* Adds MetaData, MetaDim, MetaGroup, and ArrayMetaData clases
* Updates design documentation, and adds developer guide and user guide
* Adds unit tests

Checklist
* [X] Documentation:
  * [X] Design document has been generated and added to the docs
  * [X] User's Guide has been updated
  * [X] Developer's Guide has been updated
  * [x] Documentation has been [built locally](https://e3sm-project.github.io/Omega/develop/devGuide/BuildDocs.html) and changes look as expected
* [X] Testing
  * [ ] A comment in the PR documents testing used to verify the changes including any tests that are added/modified/impacted.
  * [X] CTest unit tests for new features have been added per the approved design. 
  * [ ] Polaris tests for new features have been added per the approved design (and included in a test suite)
  * [x] Unit tests have passed. Please provide a relevant CDash build entry for verification.
  * [ ] Polaris test suite has passed
  * [ ] Performance related PRs: Please include a relevant PACE experiment link documenting performance before and after.
* [ ] Stealth Features
  * [ ] If any stealth features are included in the PR, please confirm that they have been documented.

### CODE REVIEW

File: components/omega/doc/userGuide/MetaData.md
Comment (philipwjones) [inline components/omega/doc/userGuide/MetaData.md:2]: The user guide should be targeted at those who are running the model and configuring the model for a simulation. Much of the detail currently in this doc should is more appropriate for the developer's guide which is where people who are modifying Omega code will get the necessary details for defining fields.

File: components/omega/doc/userGuide/MetaData.md
Comment (philipwjones) [inline components/omega/doc/userGuide/MetaData.md:3]: These headers are level 1 headers and need to be subheads under a main MetaData header at the top.

File: components/omega/doc/devGuide/MetaData.md
Comment (philipwjones) [inline components/omega/doc/devGuide/MetaData.md:2]: Entries for both this file and the user guide section need to be added to the top level index.html file in the doc directory

File: components/omega/doc/userGuide/MetaData.md
Comment (grnydawn) [reply components/omega/doc/userGuide/MetaData.md:2]: I moved most of the content in user guide to dev guide. In user guide, there are shorten version of it with code examples.

File: components/omega/doc/devGuide/MetaData.md
Comment (grnydawn) [reply components/omega/doc/devGuide/MetaData.md:2]: added them in index.md

File: components/omega/test/infra/MetadataTest.cpp
Comment (philipwjones) [inline components/omega/test/infra/MetadataTest.cpp:29]: @grnydawn Just one more change - in Omega, we'll be defining the metadata in various init routines and then retrieving them later in other routines. Can you create a testMetaInit routine in this test code that creates the metadata and then retrieve them later in the various other testMeta routines?  I'd like to make sure the use of shared_ptr does indeed retain the values after the initial variables go out of scope. Thanks.

Comment (grnydawn) [reply components/omega/test/infra/MetadataTest.cpp:31]: @philipwjones yes, I have added testMetaInit() and it looks working well to me.

**Review by philipwjones — APPROVED** [review]
Thanks for the changes @grnydawn  This looks fine now. Unit test passes on chrysalis/intel and the modified docs look better.

File: components/omega/src/infra/MetaData.cpp
Comment (xylar) [inline components/omega/src/infra/MetaData.cpp:121]:
```suggestion
}

int MetaData::addMetaData(
```

**Review by xylar — COMMENTED** [review]
@grnydawn, I'm sorry I didn't realize I was on the hook to review this and I wish I'd seen this sooner.  I feel like I'm holding up the process.

The code looks amazing, thanks for all the hard work!

I have a comment about the method naming convention and a few small tweaks.

I ran the CTests on Chrsyalis with Intel and they all passed.

File: components/omega/src/infra/MetaData.cpp
Comment (grnydawn) [reply components/omega/src/infra/MetaData.cpp:292]: Thanks for the comment. I will apply this in next commit.

File: components/omega/src/infra/MetaData.h
Comment (grnydawn) [reply components/omega/src/infra/MetaData.h:60]: Thanks for the comment. I will apply this in next commit.

File: components/omega/src/infra/MetaData.h
Comment (grnydawn) [reply components/omega/src/infra/MetaData.h:73]: Thanks for the comment. I will apply this in next commit.

File: components/omega/src/infra/MetaData.cpp
Comment (grnydawn) [reply components/omega/src/infra/MetaData.cpp:121]: Thanks for the comment. I will apply this in next commit.

File: components/omega/src/infra/MetaData.h
Comment (grnydawn) [reply components/omega/src/infra/MetaData.h:81]: @xylar , and @philipwjones, "*Entry()" seems to be a good idea to me. If there is no object, I will update all related method names to this style: "hasEntry()", "getEntry()", "addEntry()", and "removeEntry()".

Comment (xylar) [reply components/omega/src/infra/MetaData.h:81]: Yep, sounds good.  Let's see what @philipwjones thinks.

Comment (philipwjones) [reply components/omega/src/infra/MetaData.h:81]: I agree - let's go with Entry

Comment (grnydawn) [reply components/omega/src/infra/MetaData.h]: @xylar , @philipwjones , Great! I have just pushed the update and CI tests are passed. It seems that this PR is ready to be merged.

**Review by xylar — APPROVED** [review]
@grnydawn, looks great!  Thanks for making the changes I requested.

Ideally, someone should rerun CTests before merging.

**Comment by grnydawn** [comment]
@xylar, @philipwjones  Thanks, Unit test passes on frontier/crayclang and chrysalis/intel

> @grnydawn, looks great! Thanks for making the changes I requested.
> 
> Ideally, someone should rerun CTests before merging.

**Pull request merged**
